<h1 align="center">EDA SQL PCRF RESERVATIONS</h1>

<div style="text-align:center">
    <img src="https://www.mesaparks.com/home/showpublishedimage/16076/636493827765700000">
</div>

Now continue with the analysis, now using SQL query tools to find insights. There are some analyses that could be easier with SQL than using common Python tools. Let’s break in.

## IMPORTING LIBRARIES

We’ll import the necessary libraries for the SQL analysis:

In [1]:
import sqlite3
import pandas as pd

## DATA COLLECTION 

We can import the dataset that was filtered and used in the previous analysis:

In [2]:
data= pd.read_csv(r'C:\Users\ingde\Documents\DAP\PRCF_Reservation\data\processed\df_filtered.csv')

## SQL CONNECTION

Let's create a SQL connection, considering that Sqlite won't process over 9000 records:

In [3]:
con= sqlite3.connect('PRCF_data.db')

In [4]:
chunk_size = 9000
for i in range(0, len(data), chunk_size):
    chunk = data.iloc[i:i + chunk_size]
    if i == 0:
        chunk.to_sql('PRCF_DATA', con, if_exists='replace', index=False, method="multi")
    else:
        chunk.to_sql('PRCF_DATA', con, if_exists='append', index=False, method="multi")

In [5]:
%load_ext sql
%sql sqlite:///PRCF_data.db

The connection has been successful.

## DATA REVIEW

Let's see the data collected:

In [6]:
%sql SELECT name FROM sqlite_master WHERE type='table'

 * sqlite:///PRCF_data.db
Done.


[('PRCF_DATA',)]

In [7]:
%%sql 
SELECT * 
FROM PRCF_Data 
LIMIT 5

 * sqlite:///PRCF_data.db
Done.


[(0, 'Room - Fitness', 'Room - Fitness', -111.66297196953695, 33.432724989686506, 'Recreation Centers', 2019, 'June', 2022, 'January', '10:00:00', '2022-01-27T00:00:00.000', '11:00:00', 'Thursday', 1, 1.0, 'Reservation: Internal', 'Recreation Centers', 0, 'Mixed', 14, 1, 27, 27, 4, 6),
 (1, 'Room - Fitness', 'Room - Fitness', -111.66297196953695, 33.432724989686506, 'Recreation Centers', 2019, 'June', 2022, 'January', '11:00:00', '2022-01-27T00:00:00.000', '11:45:00', 'Thursday', 1, 0.75, 'Reservation: Internal', 'Recreation Centers', 0, 'Mixed', 14, 1, 27, 27, 4, 6),
 (2, 'Room - Fitness', 'Room - Fitness', -111.66297196953695, 33.432724989686506, 'Recreation Centers', 2019, 'June', 2022, 'January', '12:00:00', '2022-01-27T00:00:00.000', '12:45:00', 'Thursday', 1, 0.75, 'Reservation: Internal', 'Recreation Centers', 0, 'Mixed', 14, 1, 27, 27, 4, 6),
 (3, 'Room - Fitness', 'Room - Fitness', -111.66297196953695, 33.432724989686506, 'Recreation Centers', 2021, 'December', 2022, 'February', '09:30:00', '2022-02-15T00:00:00.000', '10:00:00', 'Tuesday', 30, 0.5, 'Reservation: Internal', 'Recreation Centers', 0, 'Mixed', 11, 2, 15, 15, 2, 12),
 (4, 'Gymnasium', 'Gymnasium', -111.66297196953695, 33.432724989686506, 'Recreation Centers', 2019, 'June', 2022, 'January', '11:00:00', '2022-01-27T00:00:00.000', '12:00:00', 'Thursday', 1, 1.0, 'Reservation: Internal', 'Recreation Centers', 0, 'Mixed', 14, 1, 27, 27, 4, 6)]

In [8]:
%%sql
SELECT COUNT(*)
FROM PRCF_Data

 * sqlite:///PRCF_data.db
Done.


[(434216,)]

We have to know the data types of each column for upcoming analysis:

In [9]:
cursor= con.cursor()
columnas = cursor.execute("PRAGMA table_info(PRCF_data)")

for columna in columnas:
    print(columna)

(0, 'index', 'INTEGER', 0, None, 0)
(1, 'facility_name', 'TEXT', 0, None, 0)
(2, 'facility_type', 'TEXT', 0, None, 0)
(3, 'center_longitude', 'REAL', 0, None, 0)
(4, 'center_latitude', 'REAL', 0, None, 0)
(5, 'site_name', 'TEXT', 0, None, 0)
(6, 'permit_year', 'INTEGER', 0, None, 0)
(7, 'permit_month', 'TEXT', 0, None, 0)
(8, 'event_start_year', 'INTEGER', 0, None, 0)
(9, 'event_start_month', 'TEXT', 0, None, 0)
(10, 'event_start_time', 'TEXT', 0, None, 0)
(11, 'event_end_date', 'TEXT', 0, None, 0)
(12, 'event_end_time', 'TEXT', 0, None, 0)
(13, 'day_of_week', 'TEXT', 0, None, 0)
(14, 'attendance', 'INTEGER', 0, None, 0)
(15, 'hours_reserved', 'REAL', 0, None, 0)
(16, 'schedule_type', 'TEXT', 0, None, 0)
(17, 'transaction_site', 'TEXT', 0, None, 0)
(18, 'residency_flag', 'INTEGER', 0, None, 0)
(19, 'customer_gender', 'TEXT', 0, None, 0)
(20, 'permit_hour', 'INTEGER', 0, None, 0)
(21, 'event_start_mounth_numeric', 'INTEGER', 0, None, 0)
(22, 'event_start_day_numeric', 'INTEGER', 0, None

The data agree with the information used previously.

## DATA SQL ANALYSIS

Let's see the data through SQL queries to find some business insigths:

### FACILITY COUNTING

In [10]:
%%sql
SELECT site_name, COUNT(site_name) AS Count_site_name,transaction_site
FROM PRCF_Data
GROUP BY site_name
ORDER BY Count_site_name DESC


 * sqlite:///PRCF_data.db
Done.


[('Mesa Tennis & Pickleball Center', 134189, 'Mesa Tennis & Pickleball Center'),
 ('Sports Complexes', 83158, 'Sports Complexes'),
 ('Park Sites', 81193, 'City of Mesa'),
 ('Recreation Centers', 62680, 'Recreation Centers'),
 ('Pools/Aquatics', 53043, 'City of Mesa'),
 ('Schools', 14907, 'City of Mesa'),
 ('Arizona Museum of Natural History', 2592, 'Arizona Museum of Natural History'),
 ('City of Mesa', 1371, 'City of Mesa'),
 ('i.d.e.a. Museum', 988, 'i.d.e.a. Museum'),
 ('Stadiums', 95, 'Recreation Centers')]

Easily we can see the quantity of centers dedicated to specific activities, where Tennis and other sports have the majority.

### TOP TEN FACILITIES CENTERS 

In [11]:
%%sql
SELECT facility_name, COUNT(facility_name) AS Count_facility_name, transaction_site
FROM PRCF_Data
GROUP BY facility_name
ORDER BY Count_facility_name DESC
LIMIT 10

 * sqlite:///PRCF_data.db
Done.


[('Court - Tennis', 90129, 'Mesa Tennis & Pickleball Center'),
 ('Pool', 39177, 'City of Mesa'),
 ('Field - Soccer - Complex', 32972, 'Sports Complexes'),
 ('Room - Activity', 32349, 'Recreation Centers'),
 ('Field - Softball - Complex', 25575, 'Sports Complexes'),
 ('Court - Pickleball', 25097, 'Mesa Tennis & Pickleball Center'),
 ('Field - Baseball - Complex', 22982, 'Sports Complexes'),
 ('Court - Volleyball', 19508, 'Mesa Tennis & Pickleball Center'),
 ('Ramada - Medium', 14288, 'City of Mesa'),
 ('Room - Fitness', 13890, 'Recreation Centers')]

Although Mesa Tennis & Pickleball Center has the most centers adding them up, Group Fitness has the most centers individually.

### EVENT START TIME PER DAY OF THE WEEK

In [12]:
%%sql
SELECT event_start_time, COUNT(event_start_time) AS Count_event_start_time, day_of_week, site_name
FROM PRCF_Data
GROUP BY event_start_time
HAVING Count_event_start_time > 2
ORDER BY day_of_week ASC
LIMIT 20

 * sqlite:///PRCF_data.db
Done.


[('04:00:00', 216, 'Friday', 'Pools/Aquatics'),
 ('05:00:00', 2315, 'Friday', 'Pools/Aquatics'),
 ('06:15:00', 105, 'Friday', 'Pools/Aquatics'),
 ('07:20:00', 13, 'Friday', 'Pools/Aquatics'),
 ('09:35:00', 26, 'Friday', 'Recreation Centers'),
 ('11:20:00', 359, 'Friday', 'Recreation Centers'),
 ('11:45:00', 116, 'Friday', 'Recreation Centers'),
 ('13:05:00', 50, 'Friday', 'Mesa Tennis & Pickleball Center'),
 ('13:50:00', 14, 'Friday', 'Mesa Tennis & Pickleball Center'),
 ('21:00:00', 10673, 'Friday', 'Mesa Tennis & Pickleball Center'),
 ('21:30:00', 5600, 'Friday', 'Mesa Tennis & Pickleball Center'),
 ('00:00:00', 39, 'Monday', 'Sports Complexes'),
 ('04:45:00', 36, 'Monday', 'Pools/Aquatics'),
 ('05:15:00', 3941, 'Monday', 'Pools/Aquatics'),
 ('07:30:00', 7700, 'Monday', 'Mesa Tennis & Pickleball Center'),
 ('08:30:00', 10455, 'Monday', 'Recreation Centers'),
 ('08:45:00', 856, 'Monday', 'Recreation Centers'),
 ('09:00:00', 11561, 'Monday', 'Mesa Tennis & Pickleball Center'),
 ('10:50:00', 64, 'Monday', 'Recreation Centers'),
 ('13:00:00', 10988, 'Monday', 'Mesa Tennis & Pickleball Center')]

It's rare taht Monday and Friday have events who started around at 4:00 - 5:00 hours, mostly around 5 a.m., they are days with different behaivor, but if we see the site_name columns we noticed that it's refer to pool events. In accordance with the day, Monday has a peak at 8:30, 9:00, and 13:00, it's means that the public preffer play tennis or pickball starting the week, and Friday public preffer play same sports but at night to close the week and starts the weekend.



### EVENT END TIME AND DAY OF THE WEEK

In [13]:
%%sql
SELECT event_end_time, COUNT(event_end_time) AS Count_event_end_time, day_of_week
FROM PRCF_Data
GROUP BY event_end_time
HAVING Count_event_end_time> 2
ORDER BY day_of_week DESC
LIMIT 20

 * sqlite:///PRCF_data.db
Done.


[('23:00:00', 26679, 'Wednesday'),
 ('22:00:00', 89792, 'Wednesday'),
 ('20:20:00', 3, 'Wednesday'),
 ('19:40:00', 68, 'Wednesday'),
 ('19:10:00', 8, 'Wednesday'),
 ('18:50:00', 237, 'Wednesday'),
 ('18:35:00', 65, 'Wednesday'),
 ('18:30:00', 8937, 'Wednesday'),
 ('18:15:00', 237, 'Wednesday'),
 ('17:55:00', 133, 'Wednesday'),
 ('17:50:00', 156, 'Wednesday'),
 ('16:50:00', 418, 'Wednesday'),
 ('14:55:00', 19, 'Wednesday'),
 ('14:30:00', 1868, 'Wednesday'),
 ('13:50:00', 6, 'Wednesday'),
 ('13:45:00', 119, 'Wednesday'),
 ('13:15:00', 90, 'Wednesday'),
 ('13:00:00', 11753, 'Wednesday'),
 ('12:30:00', 4716, 'Wednesday'),
 ('12:20:00', 62, 'Wednesday')]

It's clear that Wednesday has the latest event end time. Let's see the top 10 latest hours:

In [14]:
%%sql
SELECT event_end_time, COUNT(event_end_time) AS Count_event_end_time, day_of_week
FROM PRCF_Data
GROUP BY event_end_time
HAVING Count_event_end_time> 2
ORDER BY event_end_time DESC
LIMIT 10

 * sqlite:///PRCF_data.db
Done.


[('23:59:00', 76, 'Saturday'),
 ('23:45:00', 20, 'Saturday'),
 ('23:30:00', 12, 'Saturday'),
 ('23:00:00', 26679, 'Wednesday'),
 ('22:45:00', 111, 'Friday'),
 ('22:30:00', 803, 'Saturday'),
 ('22:15:00', 51, 'Thursday'),
 ('22:00:00', 89792, 'Wednesday'),
 ('21:59:00', 20, 'Friday'),
 ('21:45:00', 548, 'Friday')]

Friday and Saturday are the most common latest event end time, with Saturday having the latest ending time.

### 

### TOP 30 EVENT TIME PER DAY OF THE WEEK

In [15]:
%%sql
SELECT site_name, event_start_time, event_end_time,
CAST(event_start_time AS TIME) AS start_time,
CAST(event_end_time AS TIME) AS end_time,

CASE
WHEN (CAST(event_end_time AS TIME) - CAST(event_start_time AS TIME)) >= 0 
THEN (CAST(event_end_time AS TIME) - CAST(event_start_time AS TIME))
ELSE (CAST(event_end_time AS TIME) - CAST(event_start_time AS TIME)) + 24
END AS event_time,

COUNT(event_start_time) AS Count_event, day_of_week, transaction_site
FROM PRCF_Data
GROUP BY event_start_time, event_end_time
ORDER BY event_time DESC
LIMIT 30

 * sqlite:///PRCF_data.db
Done.


[('Recreation Centers', '00:00:00', '23:30:00', 0, 23, 23, 1, 'Saturday', 'Recreation Centers'),
 ('Recreation Centers', '00:00:00', '23:45:00', 0, 23, 23, 1, 'Saturday', 'Recreation Centers'),
 ('Pools/Aquatics', '15:30:00', '11:00:00', 15, 11, 20, 1, 'Monday', 'City of Mesa'),
 ('Sports Complexes', '04:00:00', '23:00:00', 4, 23, 19, 178, 'Saturday', 'Recreation Centers'),
 ('Recreation Centers', '05:00:00', '00:00:00', 5, 0, 19, 2, 'Friday', 'Recreation Centers'),
 ('Pools/Aquatics', '20:15:00', '15:15:00', 20, 15, 19, 5, 'Friday', 'City of Mesa'),
 ('Pools/Aquatics', '04:00:00', '22:00:00', 4, 22, 18, 24, 'Friday', 'City of Mesa'),
 ('Pools/Aquatics', '04:45:00', '22:00:00', 4, 22, 18, 12, 'Saturday', 'City of Mesa'),
 ('Sports Complexes', '05:00:00', '23:00:00', 5, 23, 18, 285, 'Thursday', 'Sports Complexes'),
 ('Recreation Centers', '05:00:00', '23:59:00', 5, 23, 18, 3, 'Friday', 'Recreation Centers'),
 ('Recreation Centers', '05:30:00', '23:00:00', 5, 23, 18, 28, 'Saturday', 'Recreation Centers'),
 ('Recreation Centers', '06:00:00', '00:00:00', 6, 0, 18, 5, 'Saturday', 'Recreation Centers'),
 ('Pools/Aquatics', '04:30:00', '21:30:00', 4, 21, 17, 4, 'Saturday', 'City of Mesa'),
 ('Pools/Aquatics', '05:00:00', '22:00:00', 5, 22, 17, 212, 'Sunday', 'City of Mesa'),
 ('Sports Complexes', '06:00:00', '23:00:00', 6, 23, 17, 6325, 'Wednesday', 'Sports Complexes'),
 ('Schools', '06:00:00', '23:59:00', 6, 23, 17, 61, 'Saturday', 'Sports Complexes'),
 ('Pools/Aquatics', '04:00:00', '20:00:00', 4, 20, 16, 1, 'Wednesday', 'City of Mesa'),
 ('Pools/Aquatics', '05:00:00', '21:45:00', 5, 21, 16, 1, 'Saturday', 'City of Mesa'),
 ('Pools/Aquatics', '05:30:00', '21:00:00', 5, 21, 16, 1, 'Wednesday', 'City of Mesa'),
 ('Pools/Aquatics', '05:30:00', '21:15:00', 5, 21, 16, 1, 'Saturday', 'City of Mesa'),
 ('Mesa Tennis & Pickleball Center', '06:00:00', '22:00:00', 6, 22, 16, 2102, 'Monday', 'Mesa Tennis & Pickleball Center'),
 ('Stadiums', '06:00:00', '22:30:00', 6, 22, 16, 1, 'Wednesday', 'Recreation Centers'),
 ('Mesa Tennis & Pickleball Center', '06:15:00', '22:00:00', 6, 22, 16, 6, 'Saturday', 'Recreation Centers'),
 ('Pools/Aquatics', '06:30:00', '22:00:00', 6, 22, 16, 175, 'Saturday', 'City of Mesa'),
 ('Pools/Aquatics', '06:30:00', '22:15:00', 6, 22, 16, 1, 'Friday', 'City of Mesa'),
 ('City of Mesa', '06:45:00', '22:00:00', 6, 22, 16, 1, 'Thursday', 'City of Mesa'),
 ('Park Sites', '07:00:00', '23:00:00', 7, 23, 16, 7298, 'Wednesday', 'Sports Complexes'),
 ('Recreation Centers', '07:00:00', '23:59:00', 7, 23, 16, 3, 'Sunday', 'Recreation Centers'),
 ('Park Sites', '07:15:00', '23:00:00', 7, 23, 16, 244, 'Thursday', 'Sports Complexes'),
 ('Pools/Aquatics', '04:00:00', '19:00:00', 4, 19, 15, 1, 'Tuesday', 'City of Mesa')]

We noticed that Recreation Centers have 23-hour reservations, normally on Saturdays. Curiously, there are reservations starting from 00:00 hours, as if there were people around the facilities at that hour. Normally, the events should start from 4:00–7:00 to 22:00–23:00; it's most probable that people start using the installations at those hours (sports mostly).

### EVENT TIME WITHOUT OCCUPATION

In [16]:
%%sql

SELECT site_name, event_start_time, event_end_time, event_time, hours_reserved_time, (hours_reserved_time - event_time) AS time_without_occupation, 
        (event_time/hours_reserved_time*100) AS occupation_percentaje, event_start_day_numeric, event_end_day_numeric, schedule_type


FROM (

        SELECT site_name, event_start_time, event_end_time, event_start_day_numeric, event_end_day_numeric, schedule_type,

            CAST(event_start_time AS TIME) AS start_time,
            CAST(event_end_time AS TIME) AS end_time,
            CAST(hours_reserved AS TIME) AS hours_reserved_time,

            CASE

                WHEN (CAST(event_end_time AS TIME) - CAST(event_start_time AS TIME)) >= 0 
                THEN (CAST(event_end_time AS TIME) - CAST(event_start_time AS TIME))
                ELSE (CAST(event_end_time AS TIME) - CAST(event_start_time AS TIME)) + 24
            
            END AS event_time

        FROM PRCF_Data

    ) AS subquery
ORDER BY hours_reserved_time DESC
LIMIT 20

 * sqlite:///PRCF_data.db
Done.


[('Sports Complexes', '00:00:00', '00:00:00', 0, 96.0, 96.0, 0.0, 5, 9, 'Reservation: Maintenance'),
 ('Sports Complexes', '00:00:00', '00:00:00', 0, 96.0, 96.0, 0.0, 5, 9, 'Reservation: Maintenance'),
 ('Sports Complexes', '00:00:00', '00:00:00', 0, 96.0, 96.0, 0.0, 5, 9, 'Reservation: Maintenance'),
 ('Stadiums', '16:00:00', '22:00:00', 6, 78.0, 72.0, 7.6923076923076925, 11, 14, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 75.0, 72.0, 4.0, 12, 15, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 75.0, 72.0, 4.0, 5, 8, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 75.0, 72.0, 4.0, 24, 27, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 75.0, 72.0, 4.0, 31, 3, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 75.0, 72.0, 4.0, 7, 10, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 75.0, 72.0, 4.0, 21, 24, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 75.0, 72.0, 4.0, 14, 17, 'Reservation: Internal'),
 ('Recreation Centers', '18:00:00', '21:00:00', 3, 51.0, 48.0, 5.88235294117647, 28, 30, 'Reservation: Internal'),
 ('Recreation Centers', '00:00:00', '23:30:00', 23, 47.5, 24.5, 48.421052631578945, 1, 2, 'Reservation: Internal'),
 ('Pools/Aquatics', '15:30:00', '11:00:00', 20, 43.5, 23.5, 45.97701149425287, 19, 21, 'Reservation: Billable'),
 ('Stadiums', '00:00:00', '00:00:00', 0, 24.0, 24.0, 0.0, 5, 6, 'Reservation: Internal'),
 ('Schools', '19:00:00', '19:00:00', 0, 24.0, 24.0, 0.0, 13, 14, 'Reservation: Billable'),
 ('Schools', '19:00:00', '19:00:00', 0, 24.0, 24.0, 0.0, 4, 5, 'Reservation: Billable'),
 ('Recreation Centers', '00:00:00', '00:00:00', 0, 24.0, 24.0, 0.0, 4, 5, 'Reservation: Internal'),
 ('Recreation Centers', '00:00:00', '23:45:00', 23, 23.75, 0.75, 96.84210526315789, 16, 16, 'Reservation: Internal'),
 ('Sports Complexes', '04:00:00', '23:00:00', 19, 19.0, 0.0, 100.0, 3, 3, 'Reservation: Maintenance')]

There are 3-4 day internal reservations, but they were used less than 8% of the time; they could be reserved for fewer hours. There are a few 24-hour billable reservations with a maximum occupancy of 48.42%; they could be reserved for 1 day.